aip/0144.md

```diff
@@ -0,0 +1,104 @@
+---
+aip:
+  id: 144
+  state: reviewing
+  created: 2020-03-19
+permalink: /144
+redirect_from:
+  - /0144
+---
+
+# Repeated fields
+
+Representing lists of data in an API is trickier than it often appears. Users
+often need to modify lists in place, and longer data series within a single
+resource pose a challenge for pagination.
+
+## Guidance
+
+Resources **may** use repeated fields where appropriate.
+
+```proto
+message Book {
+  string name = 1;
+  repeated string authors = 2;
+}
+```
+
+- Repeated fields **must** use a plural field name.
+  - If the English singular and plural words are identical ("moose", "info"),
+    the dictionary word **must** be used rather than attempting to coin a new
+    plural form.
+- Repeated fields **should** have an upper bound that will not cause a single
+  resource payload to become too large. A good rule of thumb is 100 elements.
+  - If repeated data has the chance of being too large, the API **should** use
+    a sub-resource instead.
+- Repeated fields **must not** represent the body of another resource inline.
+  Instead, the message **should** provide the resource names of the associated
+  resources.
+
+### Scalars and messages
+
+Repeated fields **should** use a scalar type (such as `string`) if they are
+certain that additional data will not be needed in the future, as using a
+message type adds significant cognitive overhead and leads to more complicated
+code.
+
+However, if additional data is likely to be needed in the future, repeated
+fields **should** use a message instead of a scalar proactively, to avoid
+parallel repeated fields.
+
+### Update strategies
+
+A resource **may** use two strategies to enable updating a repeated field:
+direct update using the [standard `Update`][aip-134] method, or custom `Add`
+and `Remove` methods.
+
+A standard `Update` method has one key limitation: the user is only able to
```

Comment:
According to our internal field masks document, a field mask **may** represent a specific field within a list, and it applies to all items within the list. I am uncertain whether this is actually done in practice and am inclined not to mention it in this document. (This would need to be sorted out in a field masks AIP.)